### PR TITLE
Add support for alternative config files with cosmiconfig

### LINKED ISSUE #118: Allow YAML or JS format for configuration file
My application is using the `@uniswap/widgets` package which for some reason does not install the necessary packages by himself and I add to install some dependencies manually.

This is my current configuration:

```json
{
  "entry": [
    "src/pages/**/*.page.{ts,tsx}",
    "src/worker/index.ts",
    "next.config.js",
    "cypress/e2e/*.cy.ts",
    "cypress/support/e2e.ts"
  ],
  "ignorePatterns": ["**/node_modules/**", "public/**"],
  "ignoreUnimported": ["src/icons/IconTemplate.js"],

  "__note": "@ethersproject/abi jotai-immer are required by @uniswap/widgets",
  "ignoreUnused": ["next", "react", "react-dom", "@ethersproject/abi", "jotai-immer"],
  "ignoreUnresolved": ["tslib", "cross-fetch", "extract-files", "form-data"]
}
```

As you can see, there is a pseudo-comment explaining why I am ignoring `@ethersproject/abi` and `jotai-immer` packages.

I would like to be able to write the configuration file as a YAML file which will allow comments:

```yaml
entry:
- src/pages/**/*.page.{ts,tsx}
- src/worker/index.ts
- next.config.js
- cypress/e2e/*.cy.ts
- cypress/support/e2e.ts
ignorePatterns:
- "**/node_modules/**"
- public/**
ignoreUnimported:
- src/icons/IconTemplate.js

ignoreUnused:
- next
- react
- react-dom
# Required by @uniswap/widgets
- "@ethersproject/abi"
- jotai-immer

ignoreUnresolved:
- tslib
- cross-fetch
- extract-files
- form-data
```

Happy to open a PR if interested.

### PR DESCRIPTION
**Pull request by ryanwilsonperkin**
Fixes #118 

Introduces `cosmiconfig` as requested in #118 in order to support loading the config from alternative file formats like:
```
.unimportedrc.js
.unimportedrc.yml
package.json > "unimported" key
```

I'm using `cosmiconfigSync` utilities instead of the async version, despite them being available since the `getConfig` is an async method.  When we use the async equivalent, we hit a segfault in Node as a result of `cosmiconfig` trying to call a dynamic import on the file within a Jest context that doesn't allow it to. Patched in a recent version of Node, and once  the test infra can require the latest version it should be fine to switch to the async version. 

See https://github.com/nodejs/node/issues/35889 and https://github.com/jestjs/jest/issues/11438

I haven't made any efforts to change the `update` function to write updates to the loaded files, as this would be difficult/impossible to update something like a .js or .yml (if using features like anchors) in a meaningful way.

A few notes on the other changes included:
1. `cosmiconfig` pulls in a newer version of TypeScript which was incompatible with the version of `@types/node` we used, updated it
2. One of the tests produced invalid JSON to a config file, which failed silently before and passed the test, but now fails loudly when `cosmiconfig` tries to read and parse the JSON. Updated it to be valid to fulfill the spirit of the test.

### CODE REVIEW

**Review by smeijer — APPROVED** [review]
Thanks! This looks amazing.

**Comment by smeijer** [comment]
@all-contributors please add @ryanwilsonperkin for code, test

File: src/__tests__/cli.ts [hunk 1]
```diff
@@ -874,7 +936,7 @@ export default promise
         { name: 'helpers/index.ts', content: '' },
         {
           name: '.unimportedrc.json',
-          content: '{ "pathTransforms": { "(\\..+)\\.js$": "$1.ts" } }',
+          content: '{ "pathTransforms": { "(\..+)\.js$": "$1.ts" } }',
```
Comment (smeijer) [inline src/__tests__/cli.ts:939]:
```suggestion
          content: '{ "pathTransforms": { "(..+).js$": "$1.ts" } }',
```

**Comment by smeijer** [comment]
I've also added you as a collaborator on the project. Please continue to make PRs as you feel the need. You can now make your branches directly on the repo rather than your fork if you want.

Thanks for your contributions! Much appreciated. :heart:

**Comment by ryanwilsonperkin** [comment]
Thanks! Had started building a similar tool to help me do some dead code analysis and then came across this tool. Happy to be able to contribute to it

**Pull request merged**